src/block/compress.rs

```diff
@@ -252,69 +242,61 @@ fn count_same_bytes(input: &[u8], candidate: usize, cur: &mut usize) -> usize {
-#[cfg(feature = "safe-encode")]
 fn write_integer(output: &mut Sink, mut n: usize) {
-    while n >= 0xFF {
-        n -= 0xFF;
-        push_byte(output, 0xFF);
+    // A predictable branch for the common case
+    if n < 0xFF {
+        push_byte(output, n as u8);
+        return;
     }
 
-    // Write the remaining byte.
-    push_byte(output, n as u8);
-}
-
-/// Write an integer to the output in LSIC format.
-#[inline]
-#[cfg(not(feature = "safe-encode"))]
-fn write_integer(output: &mut Sink, mut n: usize) {
-    // Write the 0xFF bytes as long as the integer is higher than said value.
-    push_u32(output, 0xFFFFFFFF);
+    // Write the 0xFF bytes in chunks of 4.
     while n >= 4 * 0xFF {
         n -= 4 * 0xFF;
-        output.set_pos(output.pos() + 4);
         push_u32(output, 0xFFFFFFFF);
     }
 
-    // Updating output len for the remainder
-    output.set_pos(output.pos() + 1 + n / 255);
-    unsafe {
-        // Write the remaining byte.
-        *output.as_mut_ptr().sub(1) = (n % 255) as u8;
-    };
-}
-
-#[inline]
-#[cfg(not(feature = "safe-encode"))]
-fn push_u32(output: &mut Sink, el: u32) {
-    unsafe {
-        let out_ptr = output.as_mut_ptr();
-        core::ptr::copy_nonoverlapping(el.to_le_bytes().as_ptr(), out_ptr, 4);
-    }
+    // Write the last 0..=3 0xFF bytes + remainder.
+    let ffs = n / 0xFF;
+    let rem = n % 0xFF;
+    push_u32(
+        output,
+        0xFFFFFFFF ^ ((!(rem as u8) as u32) << ((3 - ffs) * 8)).to_be(),
+    );
+    output.pos -= 3 - ffs as usize;
 }
 
 /// Handle the last bytes from the input as literals
 #[cold]
-fn handle_last_literals(output: &mut Sink, input: &[u8], input_size: usize, start: usize) -> usize {
-    let lit_len = input_size - start;
+fn handle_last_literals(output: &mut Sink, input: &[u8], start: usize) -> usize {
+    let lit_len = input.len() - start;
 
     let token = token_from_literal(lit_len);
     push_byte(output, token);
-    // output.push(token);
     if lit_len >= 0xF {
         write_integer(output, lit_len - 0xF);
     }
     // Now, write the actual literals.
-    copy_literals(output, &input[start..]);
+    output.extend_from_slice(&input[start..]);
     output.pos()
 }
 
 /// Moves the cursors back as long as the bytes match, to find additional bytes in a duplicate
 ///
 #[inline]
 #[cfg(feature = "safe-encode")]
-pub fn backtrack_match(candidate: &mut usize, cur: &mut usize, literal_start: usize, input: &[u8]) {
-    // TODO: It should be possible remove all bounds checks, since we are walking backwards
-    while *candidate > 0 && *cur > literal_start && input[*cur - 1] == input[*candidate - 1] {
+pub fn backtrack_match(
+    input: &[u8],
+    cur: &mut usize,
+    literal_start: usize,
+    source: &[u8],
+    candidate: &mut usize,
+) {
+    let left = input[literal_start..*cur].iter().rev().copied();
+    let right = source[..*candidate].iter().rev().copied();
+    for (a, b) in left.zip(right) {
+        if a != b {
+            break;
+        }
```

Comment:
Did you benchmark this? I tried a similar solution which was slower

Reply:
I think I did, but I'll double check :+1: